src/Encoder/Encoder.php

```diff
@@ -83,12 +84,16 @@ public static function encode(
             + $dataBits->getSize();
         $provisionalVersion = self::chooseVersion($provisionalBitsNeeded, $ecLevel);
 
-        // Use that guess to calculate the right version. I am still not sure
-        // this works in 100% of cases.
-        $bitsNeeded = $headerBits->getSize()
```

Comment:
I'd argue that the code should still make its own calculation and thus verify that the given version number full fills the requirement (i. e. that it is not lower than the minimal required version number; Otherwise throw an exception.

Reply:
Developers documentation (in Slovenian): https://upn-qr.si/uploads/files/NavodilaZaProgramerjeUPNQR.pdf on a page 10 is input example

Current source code will calculate Version(10)
``` php
$renderer = new \BaconQrCode\Renderer\ImageRenderer(
            new \BaconQrCode\Renderer\RendererStyle\RendererStyle(256),
            new \BaconQrCode\Renderer\Image\SvgImageBackEnd()
);
$writer = new \BaconQrCode\Writer($renderer);
$output = $writer->writeString(
    "UPNQR\n\n\n\n\nJanez Novak\nDunajska ulica 1\n1000 Ljubljana\n" .
        "00000008105\n\n\nRENT\nPlačilo najemnine za marec 2017\n01.04.2017\n" .
        "SI56020170014356205\nSI121234567890120\nRentaCar d.o.o.\nPohorska ulica 22\n" .
        "2000 Maribor\n201\n",
    'ISO-8859-2',
    \BaconQrCode\Common\ErrorCorrectionLevel::forBits(0)
);
```

So, changes in PR will use the correct Version(15)
``` php
$output = $writer->writeString(
    "UPNQR\n\n\n\n\nJanez Novak\nDunajska ulica 1\n1000 Ljubljana\n" .
        "00000008105\n\n\nRENT\nPlačilo najemnine za marec 2017\n01.04.2017\n" .
        "SI56020170014356205\nSI121234567890120\nRentaCar d.o.o.\nPohorska ulica 22\n" .
        "2000 Maribor\n201\n",
    'ISO-8859-2',
    \BaconQrCode\Common\ErrorCorrectionLevel::forBits(0),
    \BaconQrCode\Common\Version::getVersionForNumber(15)
);
```

If I understand instead adding `if version set then do ... ` new code should look like (see below). But this will always thrown an exception and never create a QR code.
``` php
$version = self::chooseVersion($bitsNeeded, $ecLevel);

if (null !== $forcedVersion
  && $version->getVersionNumber() < $forcedVersion->getVersionNumber()
) {
  throw new \Exception();
}
```

Reply:
Something like that, yes :)

Reply:
Sorry for a very late response.

If script throws Exception, i will never get a requried/forced QR code Version. So, why would i use it if it does not work. Current `choseVersion` is returning version: 10, 11, 12, 13, 14, 15 - i did not check futher more.

There is another option, but it's not a perfect solution. So, the code would look like something like this:

```php
$testVersion = 1;
do {
    // Current code to get version: lined 83-95
   $version = self::chooseVersion($bitsNeeded, $ecLevel, $testVersion);
   $testVersion = $version->getVersionNumber() +1;
} while (null !== $forcedVersion && $version->getVersionNumber() < $forcedVersion->getVersionNumber());

if (null !== $forcedVersion && $version->getVersionNumber() != $forcedVersion->getVersionNumber()) {
    throw new WriterException('Invalid QR version number');
}
...
// Additional param int $versionNum = 1;
private static function chooseVersion(int $numInputBits, ErrorCorrectionLevel $ecLevel, int $versionNum = 1) : Version
{
    for ($versionNum; $versionNum <= 40; ++$versionNum) {
...
```